### PR TITLE
feat(sdk): auto-derive fee token from warp route token type

### PR DESCRIPTION
**Pull request by paulbalaji**
## Summary

Simplifies fee token configuration by removing the `token` field from fee input schemas. The SDK now automatically derives the correct fee token address based on the warp route token type:

- **Native tokens** → `AddressZero`
- **Collateral tokens** → collateral token address (`config.token`)
- **Synthetic tokens** → router address (the HypERC20 itself)

This eliminates user error and simplifies warp route configs - users no longer need to specify (or know) the fee token address.

### Fee Input Schemas (`fee/types.ts`)
- Removed `token` from `BaseFeeConfigInputSchema` - users no longer specify fee token
- Added proper input schemas for `ProgressiveFee` and `RegressiveFee`
- `RoutingFeeInputConfigSchema` no longer includes `maxFee`/`halfAmount` (contract hardcodes to `type(uint256).max`)
- Added validation to reject `bps: 0` (prevents division by zero in `convertFromBps`)

### Fee Token Resolution (`token/configUtils.ts`)
- Added `resolveTokenFeeAddress()` utility that determines fee token based on token type
- Called at deploy time after router addresses are known
- Recursively resolves tokens for nested `RoutingFee` configs

### Fee Parameter Derivation (`fee/EvmTokenFeeReader.ts`)
- Simplified `convertFromBps()` to use constant divisor (`10^36`) instead of `totalSupply`
- **Fixes division by zero** for newly deployed synthetic tokens with zero supply
- Ensures no overflow in `LinearFee._quoteTransfer()` for any realistic transfer amount

### Schema Validation (`token/types.ts`)
- Removed token matching validation (no longer needed since SDK derives token)
- Simplified `populateFeeOwner` to only handle owner population

## Example

Before (user had to specify token):
```yaml
tokenFee:
  type: LinearFee
  token: "0x..."  # Had to know this ahead of time
  bps: 100
```

After (SDK derives token automatically):
```yaml
tokenFee:
  type: LinearFee
  bps: 100
```

## Testing

- Unit tests for `resolveTokenFeeAddress()` covering all token types (native, collateral, synthetic)
- Unit tests for `LinearFeeInputConfigSchema` validation (bps-only, maxFee/halfAmount-only, edge cases)
- Hardhat tests for `EvmTokenFeeModule.expandConfig()` with zero-supply tokens
- CLI e2e tests for synthetic token with LinearFee (fee token resolved to router)
- CLI e2e tests for collateral token with explicit maxFee/halfAmount

## Summary by CodeRabbit

* **New Features**
  * Fee tokens are auto-derived at deploy time based on token type; specifying fee token is often unnecessary.
  * Public export added for linear fee configs for easier access.

* **Improvements**
  * Accept either bps or explicit maxFee/halfAmount; missing values are auto-computed.
  * Nested routing fees inherit tokens automatically; zero-supply edge cases handled more robustly.
  * Validation flow simplified for token-fee population.

* **Tests**
  * Expanded unit and end-to-end tests covering these scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: typescript/sdk/src/fee/types.ts [hunk 1]
```diff
@@ -42,12 +42,19 @@ export const onChainTypeToTokenFeeTypeMap: Record<
...
 
+// For deployed/read configs - token is required
 export const BaseFeeConfigSchema = z.object({
   token: ZHash,
   owner: ZHash,
 });
 export type BaseTokenFeeConfig = z.infer<typeof BaseFeeConfigSchema>;
 
+// For input configs - token is optional (resolved at deploy time for synthetic tokens)
+export const BaseFeeConfigInputSchema = z.object({
```
Comment (paulbalaji) [inline typescript/sdk/src/fee/types.ts]: worth updating progressive/regressive schemas as well to separate config vs input config?

File: typescript/sdk/src/token/EvmERC20WarpModule.ts [hunk 2]
```diff
@@ -1020,6 +1026,31 @@ export class EvmERC20WarpModule extends HyperlaneModule<
...
+    const resolved: TokenFeeConfigInput = { ...feeConfig };
+
+    if (feeConfig.token === undefined) {
+      resolved.token = routerAddress;
+    }
+
+    if (feeConfig.type === TokenFeeType.RoutingFee) {
+      const routingConfig = feeConfig as RoutingFeeInputConfig;
+      if (routingConfig.feeContracts) {
+        (resolved as RoutingFeeInputConfig).feeContracts = Object.fromEntries(
```
Comment (ltyu) [inline typescript/sdk/src/token/EvmERC20WarpModule.ts]: can we use objMap here?

File: typescript/sdk/src/token/EvmERC20WarpModule.ts [hunk 3]
```diff
@@ -1020,6 +1026,31 @@ export class EvmERC20WarpModule extends HyperlaneModule<
...
 
+  private resolveTokenFeeAddress(
+    feeConfig: TokenFeeConfigInput,
+    routerAddress: Address,
+  ): TokenFeeConfigInput {
+    const resolved: TokenFeeConfigInput = { ...feeConfig };
+
+    if (feeConfig.token === undefined) {
+      resolved.token = routerAddress;
+    }
```
Comment (xeno097) [inline typescript/sdk/src/token/EvmERC20WarpModule.ts]: Shouldn't we do this only when the token is a synthetic? We shouldn't rely on the caller having validated the input

File: typescript/sdk/src/fee/types.ts [hunk 4]
```diff
@@ -42,12 +42,19 @@ export const onChainTypeToTokenFeeTypeMap: Record<
...
+  token: ZHash.optional(),
```
[the remark this answers is not in the document]
Comment (ltyu) [reply typescript/sdk/src/fee/types.ts]: >Are there any cases where we would like to have fees paid in another token that is not the collateral of the current warp contract

dont think you can because fee must match token

File: typescript/cli/src/tests/ethereum/warp/warp-deploy-2.e2e-test.ts [hunk 5]
```diff
@@ -393,6 +396,46 @@ describe('hyperlane warp deploy e2e tests', async function () {
...
+        await readWarpConfig(
+          CHAIN_NAME_3,
+          COMBINED_WARP_CORE_CONFIG_PATH,
+          WARP_DEPLOY_OUTPUT_PATH,
+        )
+      )[CHAIN_NAME_3];
+
+      expect(syntheticConfig.tokenFee).to.exist;
+      expect(syntheticConfig.tokenFee?.type).to.equal(TokenFeeType.LinearFee);
+      expect((syntheticConfig.tokenFee as TokenFeeConfig | undefined)?.token).to
```
Comment (antigremlin) [inline typescript/cli/src/tests/ethereum/warp/warp-deploy-2.e2e-test.ts]: Should we check that it's equal to the expected token router?

File: typescript/sdk/src/fee/EvmTokenFeeModule.ts [hunk 6]
```diff
@@ -112,63 +136,91 @@ export class EvmTokenFeeModule extends HyperlaneModule<
...
         ? await promiseObjAll(
-            objMap(config.feeContracts, async (_, innerConfig) => {
-              return EvmTokenFeeModule.expandConfig({
-                config: innerConfig,
-                multiProvider,
-                chainName,
-              });
-            }),
+            objMap(
+              inputFeeContracts as Record<string, ResolvedTokenFeeConfigInput>,
```
Comment (xeno097) [inline typescript/sdk/src/fee/EvmTokenFeeModule.ts]: ditto type assertion
